File: leetcode/easy/53-max-subarr/kadane.py
"""
  @author Vortexx2

  Runtime - 56 ms
  Memory Usage - 15.1 MB
"""
from typing import List


class Solution:
  def maxSubArray(self, nums: List[int]) -> int:
    max_sum = nums[0]
    curr_sum = nums[0]

    for i in range(1, len(nums)):

      temp = curr_sum + nums[i]

      if nums[i] > temp:
        curr_sum = nums[i]

      else:
        curr_sum = temp

      max_sum = max(max_sum, curr_sum)

    return max_sum


if __name__ == '__main__':
  sol = Solution()
  print(sol.maxSubArray([5, 4, -1, 7, 8]))

File: leetcode/medium/53-max-subarr/kadane.cpp
/**
 * @file kadane.cpp
 * @author Vortexx2
 * @brief 
 * @date 14-07-2021
 * 
 * Runtime - 0ms 
 * Memory Usage - 13.1 MB
 */
#include <iostream>
#include <vector>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>
#include <climits>

using namespace std;

class Solution
{
public:
  int maxSubArray(vector<int> &nums)
  {
    int maxSum = INT_MIN;
    int currSum = 0;
    int temp = 0;

    for (int i = 0; i < nums.size(); i++)
    {
      temp = currSum + nums[i];

      if (nums[i] > temp)
        currSum = nums[i];

      else
        currSum = temp;

      if (currSum > maxSum)
      {
        maxSum = currSum;
      }
    }

    return maxSum;
  }
};

int main()
{
  cout << INT_MIN << endl;
}
